File: LeetCode/209/main.cpp
class Solution {
public:
  int minSubArrayLen(int target, vector<int> &arr) {

    int n = arr.size();
    int left = 0;
    int ans = INT_MAX;
    int s = 0;

    for (int i = 0; i < n; i++) {
      if (arr[i] == target) {
        return 1;
      }
      s += arr[i];
      while (s >= target) {
        ans = min(ans, i - left + 1);
        s -= arr[left];
        left++;
      }
    }

    return (ans == INT_MAX) ? 0 : ans;
  }
};

File: LeetCode/209/main.py
"""
DIFFICULTY : medium
TAGS : array, binary search, sliding window, prefix sum
"""


class Solution:
    def minSubArrayLen(self, target: int, arr: List[int]) -> int:
        n = len(arr)
        left = 0
        ans = float("inf")
        s = 0
        for i in range(n):
            s += arr[i]
            while s >= target:
                ans = min(ans, i - left + 1)
                s -= arr[left]
                left += 1

        if ans == float("inf"):
            return 0
        return ans
